### DISCUSSION TITLE
Parsing Error Correcting Code as Data (?)

### DISCUSSION THREAD

**Discussion by Meerkov** [vgmtrans/vgmtrans#290] [General]
**Describe the bug**
When parsing music files from 3Xtreme (PSX), a strange result occurs. If you parse an .img, no Instrument Sets will successfully extract... However, if you extract the same file using a separate program (e.g. jpsxdec) and then run the resulting FATHEAD.SND through VGMTrans, then you *will* successfully get those Instrument Sets... Additionally, the KeyHigh and KeyLow values don't appear to match. Indeed, even though you use the same original source for both files, VGMTrans will get a corrupted file.

**To Reproduce**
Setup:
0. Obtain JPSXDec, a copy of any PS1 game that doesn't successfully parse instrument sets (in my case, 3Xtreme is a good example), VGM Trans, and DiscoHawk (part of the BizHawk suite).
1. First, ensure you have the game in .img format by running 3Xtreme through DiscoHawk. This results in 3Xtreme_hawked.img (ensuring it's a single file, and not split into tracks, just in case).

First, obtain the relevant file, FATHEAD.SND for comparison
2. Run 3Xtreme_hawked.img through JPSXDec to perform file extraction. Find the file FATHEAD.SND and extract it using **NORMAL** mode.
3. Open FATHEAD.SND in a hexeditor for comparison purposes later (e.g. Hexed.it )

Next, let's do a successful workflow:
4. Run VGMTrans in debug mode through visual studio,
5. Set a Breakpoint in Vab::GetHeaderInfo().
6. Set a Breakpoint in VagRgn:LoadRgn() at this line:
`    message << L"Low Key (" << keyLow << L") is higher than High Key (" << keyHigh << L")  Offset: 0x" << std::hex << dwOffset;`
7. Load FATHEAD.SND (drag and drop) into VGMTrans
8. Notice that the breakpoint in GetHeaderInfo is hit.
9. Press F5 twice, and then see that the following has loaded (Pay special attention to the two VABs)
`Three PS1 SEQ, two VAB, and one VGMSampColl.`
10. Compare the VAB header to the file FATHEAD.SND. Notice it has 14 instruments, among other details of the header. This means we've successfully parsed instrument sets deriving from the original 3Xtreme img! This means the file is not corrupted, and music can be extracted without errors, using this workflow.

Now, let's show the failing result
11. Try to load 3Xtreme_hawked.img in VGMTrans
12. Turn off the breakpoint set in step (6), and press F5 one time. This will bring you to the second VH file.
13. Turn the breakpoint in step 6 back on.
14. You can do a line-by-line step and confirm that the second file is indeed FATHEAD.SND (14 instruments, the other header information matches, etc).
15. Either step through manually, or press F5 to quickly reach the breakpoint that was set in step 6. Notice that FATHEAD.SND has not finished parsing yet, but you are hitting an error that you did not hit on the previous attempt...
16. In fact, all of the music files will hit this error message, and *no* instrument sets will successfully parse!
17. This implies that the values that it is reading are somehow *not* matching....

Check the PS1 Seq Files for comparison
18. Check the first PS1 SEQ from FATHEAD.SND. Scroll down to line `00004FFB`, and see the ascii `..8...9....y....`
19. Check the *second* PS1 SEQ from 3Xtreme_hawked.img. Scroll down the same number of lines (count if you want to, or use addition).
20. Notice that the file instead reads `..8....Tp..ê....`
21. This is the end of the second file, but the first file continues for a long long time further. But they should be the *same file* (indeed, they are byte-for-byte identical up until this point).

Summary:
Something really fishy is going on in the file parsing here. One file parsed standalone works seemingly flawlessly. However, parsed directly from the original source that created that working file, results in a corrupted file...

**Expected behavior**
The two files would be identical.

**Additional context**
Given that I'm working on debugging this section of parser for other issues, I made sure to run this on the Master branch, and this indeed reproduces.

**Comment by Meerkov** [comment]
I looked a little bit more...

One option when extracting the file from JSPXDec is whether to extract as a RAW (2352 byte) or normal (2048 byte) sector.

The above reproduction steps utilize **NORMAL** extraction, resulting in a FATHEAD.SND that is approximately 520kb.

However, if you extract a **RAW** FATHEAD.SND, that is approximately 590kb, you will get the broken parsing that VGMTrans results in... So, what does this mean?

Well, this means to me that there seems to be a fundamental flaw in VGMTrans' handling of raw sectors. Clearly, the file begins parsing alligned, but there must be an an error correcting code that begins around line `00004FFB`. VGM Trans is not properly skipping the ECC, and instead interpreting this as data...

That's currently my best guess for what the issue is.

**Comment by Meerkov** [comment]
[3Xtreme (USA).zip](https://github.com/vgmtrans/vgmtrans/files/7936963/3Xtreme.USA.zip)
Here are the two files (RAW and NORMAL) of the FATHEAD.SND that you can use to reproduce.

**Comment by sykhro** [comment]
VGMTrans has no idea about sectors - the whole file is scanned looking for the magic `pBAV` and parsing starts from there, reading sequences of bytes (0x20 bytes for the header, 0x200 [iirc] per instruments, ...), with no alignment assumptions. So, if there's ECC data after the VAB header, it's interpreted as VAB data. I can't think of a way to solve this besides some sort of "somehow strip raw sector info and try again". And somehow telling the user that BIN/CUE should not be used?

**Comment by Meerkov** [comment]
That might explain why I was having such difficulty getting anything to work right. Considering PS1 games will often use ECC on data (and I think these files would count as Data rather than CD audio), I imagine this likely affects 90%+ of games...

I'm not really sure what the right outcome is here: Is it even plausible for VGM Trans to handle RAW format? Does it make sense to treat removing the ECC as the default behavior when loading .img files?

Edit: Would running the .img through another program to strip the ECC even result in correct output, given that some files are going to be CD Audio with no ECC?

**Comment by sykhro** [comment]
> I'm not really sure what the right outcome is here: Is it even plausible for VGM Trans to handle RAW format? Does it make sense to treat removing the ECC as the default behavior when loading .img files?

When a file is put into VGMTrans, we throw every scanner for music data we have at it. From the pov of the program, FATHEAD.SND could contain nds, snes, ps2, sega saturn, cps, [...] music data. It works on individual files (ripped from games) and memory dumps. If the data is in some sort of container like minipsf, we can extract it (see the Loader classes). 

Given the architecture, the thing that could work would be to implement a Loader that reads the extension and interprets the file as a CD/DVD dump if it has the .img extension (assuming there are no other consoles around that use `.img` as the standard extension for their raw non-CD dumps!). Given that there's already plenty of tools around that do this, I'm not sure if it's worth the effort.. it all boils down to the UX of feedback/what the program supports, again (argh)

**Comment by loveemu** [comment]
As sykhro mentioned, this is currently a specification. I feel that this issue should be labeled/titled as a feature request. Or it could be closed as a question.

**Reply by sykhro** [reply]
Moved to discussions

**Reply by Meerkov** [reply]
I think it makes sense to treat this as a feature request. I found a way to work around this issue under at least some circumstances, (future people can find it in bug #251). Native support without a separate tool should be treated as a feature.

Especially now that more errors will be logged, it can provide users a "hint" if so desired that they should look at this as being the issue they are hitting. (Messages along the lines of "too many notes" or "highkey < lowkey" indicate a RAW format input).

**Reply by sykhro** [reply]
> Native support without a separate tool should be treated as a feature.

There isn't a way to detect if a file is a CD/DVD dump with ECC data or not. The best we could do is to add a dedicated "Load raw CD/DVD image" action in the menu or somewhere else.

**Reply by Meerkov** [reply]
Correct me if I'm wrong, because I'm still pretty new to this subject... but can't you just run the ECC algorithm to determine if the 256 bytes following the data block is an ECC signature? You could check if there was 0 or 1 errors (I'm not sure which ECC algorithm is used, and the maximum number of errors it can correct for, but I'll assume 1 error here).

That should be as easy as just running the ECC algorithm on each 2048 block using the next 256 bytes, right? And then, assuming it succeeds, it could hand over the scan to the other relevant PS1 data format scanners.

**Reply by sykhro** [reply]
> That should be as easy as just running the ECC algorithm on each 2048 block using the next 256 bytes, right? And then, assuming it succeeds, it could hand over the scan to the other relevant PS1 data format scanners.

Or, given that the user can tell with 100% confidence that they have a raw disk dump, have them press "Load RAW CD image".
I don't think anyone would be willing to implement cross-interleaved Reed-Solomon* just to save a click.

*DVDs uses yet another scheme and it would be needed for PS2 roms, which I'm guessing are plagued by the same issue

**Reply by Meerkov** [reply]
Well ok that's fair. But I guess I was just confirming that such a method *would* work. And there are libraries that do reed-solomon in C++, but I think the effort is a valid point. Unless there are other reasons to pull in such libraries, it might be better just to have a user setting, disambiguation dialog, or like you suggest, make it not work on click/drag and require a different menu entry.
